src/utils/event.ts

```diff
@@ -0,0 +1,43 @@
+import { IBlockStub, ApplicationError } from "../dataEntities";
+
+/**
+ * This is a generic class to use as base for concrete classes to manage events and event listeners.
+ * The type parmeter TListener is the type of the event listeners, which should be a function returning a Promise.
+ * Concrete subclasses must implement an `emit` method method to actually emits the event and calls all the listeners,
+ * awaiting each one of them.
+ */
+export abstract class Event<TListener extends Function> {
```

Comment:
Cant you enforce that Function returns a promise by extending from
```
TListener extends () => Promise<any>
```
or
```
TListener extends () => Promise<TReturn>
```
if you want to consistent return type

Reply:
I didn't know you can constrain versus arrow types like that.
That doesn't work yet, as subclasses (like BlockEvent) might take listeners with a parameter. This seems to work:
```
export abstract class Event<TListener extends (...args: any[]) => Promise<void>> {
```

where I also constrained the return value to `void`, no reason to expect anything else.

Reply:
Changed in 749e18e.